File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/java/JdkResolver.java
package org.jetbrains.bsp.bazel.server.sync.languages.java;

import io.vavr.PartialFunction;
import io.vavr.collection.List;
import io.vavr.collection.Seq;
import io.vavr.control.Option;
import java.net.URI;
import java.util.Objects;
import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo;
import org.jetbrains.bsp.bazel.server.sync.BazelPathsResolver;

public class JdkResolver {
  private final BazelPathsResolver bazelPathsResolver;

  public JdkResolver(BazelPathsResolver bazelPathsResolver) {
    this.bazelPathsResolver = bazelPathsResolver;
  }

  public Option<Jdk> resolve(Seq<TargetInfo> targets) {
    var allCandidates = targets.flatMap(this::resolveJdk);
    if (allCandidates.isEmpty()) return Option.none();
    var latestVersion = candidatesWithLatestVersion(allCandidates);
    var complete = allCandidates.filter(JdkCandidate::isComplete);
    var latestVersionAndComplete = latestVersion.filter(JdkCandidate::isComplete);
    return Option.<JdkCandidate>none()
        .orElse(() -> pickCandidateFromJvmRuntime(latestVersionAndComplete))
        .orElse(() -> pickAnyCandidate(latestVersionAndComplete))
        .orElse(() -> pickCandidateFromJvmRuntime(complete))
        .orElse(() -> pickAnyCandidate(complete))
        .orElse(() -> createHybridCandidate(allCandidates))
        .orElse(() -> pickAnyCandidate(allCandidates))
        .flatMap(JdkCandidate::asJdk);
  }

  private Seq<JdkCandidate> candidatesWithLatestVersion(Seq<JdkCandidate> candidates) {
    var latestVersion = findLatestVersion(candidates);
    var byVersion =
        latestVersion.map(
            version -> candidates.filter(candidate -> candidate.version.contains(version)));
    // sorted for deterministic output
    return byVersion.getOrElse(List.empty()).sortBy(c -> c.javaHome.toString());
  }

  private Option<String> findLatestVersion(Seq<JdkCandidate> candidates) {
    return candidates.flatMap(c -> c.version).maxBy(s -> Integer.valueOf(s));
  }

  private Option<JdkCandidate> pickCandidateFromJvmRuntime(Seq<JdkCandidate> candidates) {
    return candidates.find(candidate -> candidate.isRuntime);
  }

  private Option<JdkCandidate> pickAnyCandidate(Seq<JdkCandidate> candidates) {
    return candidates.headOption();
  }

  private Option<JdkCandidate> createHybridCandidate(Seq<JdkCandidate> candidates) {
    var version = findLatestVersion(candidates);
    var javaHome =
        candidates
            .sortBy(c -> !c.isRuntime)
            .collect(PartialFunction.unlift(c -> c.javaHome))
            .headOption();
    return javaHome.map(jh -> new JdkCandidate(true, javaHome, version));
  }

  private Option<JdkCandidate> resolveJdk(TargetInfo targetInfo) {
    var hasRuntimeJavaHome =
        targetInfo.hasJavaRuntimeInfo() && targetInfo.getJavaRuntimeInfo().hasJavaHome();

    var hasToolchainJavaHome =
        targetInfo.hasJavaToolchainInfo() && targetInfo.getJavaToolchainInfo().hasJavaHome();

    var javaHomeFile =
        hasRuntimeJavaHome
            ? (targetInfo.getJavaRuntimeInfo().getJavaHome())
            : (hasToolchainJavaHome ? targetInfo.getJavaToolchainInfo().getJavaHome() : null);

    var javaHome = Option.of(javaHomeFile).map(bazelPathsResolver::resolveUri);

    Option<String> version =
        targetInfo.hasJavaToolchainInfo()
            ? Option.some(targetInfo.getJavaToolchainInfo().getSourceVersion())
            : Option.none();

    return Option.some(new JdkCandidate(hasRuntimeJavaHome, javaHome, version));
  }

  private static class JdkCandidate {
    final boolean isRuntime;
    final Option<URI> javaHome;
    final Option<String> version;

    public JdkCandidate(boolean isRuntime, Option<URI> javaHome, Option<String> version) {
      this.isRuntime = isRuntime;
      this.javaHome = javaHome;
      this.version = version;
    }

    public boolean isComplete() {
      return javaHome.isDefined() && version.isDefined();
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      JdkCandidate that = (JdkCandidate) o;
      return isRuntime == that.isRuntime
          && javaHome.equals(that.javaHome)
          && version.equals(that.version);
    }

    @Override
    public int hashCode() {
      return Objects.hash(isRuntime, javaHome, version);
    }

    public Option<Jdk> asJdk() {
      return version.map(ver -> new Jdk(ver, javaHome));
    }
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/java/JdkResolver.kt
package org.jetbrains.bsp.bazel.server.sync.languages.java

import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo
import org.jetbrains.bsp.bazel.server.sync.BazelPathsResolver
import java.net.URI
import kotlin.io.path.toPath

class JdkResolver(
    private val bazelPathsResolver: BazelPathsResolver,
    private val jdkVersionResolver: JdkVersionResolver
) {

  fun resolve(targets: Sequence<TargetInfo>): Jdk? {
    val allCandidates = targets.mapNotNull { resolveJdkData(it) }.sortByFrequency().map { JdkCandidate(it) }
    if (allCandidates.none()) return null
    val latestVersion = candidatesWithLatestVersion(allCandidates)
    val complete = allCandidates.filter { it.isComplete }
    val latestVersionAndComplete = latestVersion.filter { it.isComplete }
    return (
        pickCandidateFromJvmRuntime(latestVersionAndComplete) ?:
        pickAnyCandidate(latestVersionAndComplete) ?:
        pickCandidateFromJvmRuntime(complete) ?:
        pickAnyCandidate(complete) ?:
        pickAnyCandidate(allCandidates)
      )?.asJdk()
  }

  fun resolveJdk(target: TargetInfo): Jdk? {
    return resolveJdkData(target)?.let { JdkCandidate(it).asJdk() }
  }

  private fun candidatesWithLatestVersion(candidates: Sequence<JdkCandidate>): Sequence<JdkCandidate> =
      findLatestVersion(candidates)
        ?.let { version -> candidates.filter { it.version == version } }
        .orEmpty()

  private fun findLatestVersion(candidates: Sequence<JdkCandidate>): String? =
      candidates.mapNotNull { it.version }.maxByOrNull { Integer.parseInt(it) }

  private fun pickCandidateFromJvmRuntime(candidates: Sequence<JdkCandidate>) =
      candidates.find { it.isRuntime }

  private fun pickAnyCandidate(candidates: Sequence<JdkCandidate>): JdkCandidate? =
      candidates.firstOrNull()

  private fun resolveJdkData(targetInfo: TargetInfo): JdkCandidateData? {
    val hasRuntimeJavaHome = targetInfo.hasJavaRuntimeInfo() &&
        targetInfo.javaRuntimeInfo.hasJavaHome()
    val hasToolchainJavaHome = targetInfo.hasJavaToolchainInfo() &&
        targetInfo.javaToolchainInfo.hasJavaHome()

    val javaHomeFile =
        if (hasRuntimeJavaHome)
          targetInfo.javaRuntimeInfo.javaHome
        else if (hasToolchainJavaHome)
          targetInfo.javaToolchainInfo.javaHome
        else
          null
    val javaHome = javaHomeFile?.let { bazelPathsResolver.resolveUri(it) }

    return JdkCandidateData(hasRuntimeJavaHome, javaHome)
        .takeIf { javaHome != null }
  }

  private inner class JdkCandidate(private val data: JdkCandidateData) {
    val version = data.javaHome?.let { jdkVersionResolver.resolve(it.toPath()) }?.toString()
    val javaHome by data::javaHome
    val isRuntime by data::isRuntime
    val isComplete = javaHome != null && version != null
    fun asJdk(): Jdk? = version?.let { Jdk(it, javaHome) }
  }

  private data class JdkCandidateData(
      val isRuntime: Boolean,
      val javaHome: URI?,
  )

private fun <A> Sequence<A>.sortByFrequency(): Sequence<A> =
      groupBy { it }.values
        .sortedByDescending { it.size }
        .map { it.first() }
        .asSequence()

}
